### DISCUSSION TITLE
Avoid of `import.meta.env` usage in code

### DISCUSSION THREAD

**Discussion by maxsbelt** [pmndrs/jotai#1974] [General] [answered]
I'm using `vite` as build system and after `jotai` installation I noticed interesting issue. After build code chunk with `jotai` dependency  contains multiple objects from `import.meta.env`. Example:

<img width="1390" alt="Screenshot 2023-06-03 at 20 38 22" src="https://github.com/pmndrs/jotai/assets/950216/8eb2aebf-f0bb-4f4e-babb-58c203e6b819">

I think that happens because code like:
```
import.meta.env?.MODE !== 'production'
```

Is transformed to:

```
import.meta.env && import.meta.env.MODE !== 'production'
```

And `vite` replace it with:

```
{ ACCESS_TOKEN: 'xxx', PROD: true, ... }&&"production"!=="production"
```

I think it can be problematic because if this library be extracted to `vendors` chunk it will invalidate this chunk on each env variables update which is bit unexpected. Can this logic be replaced with [more standard `process.env.NODE_ENV`](https://github.com/vitejs/vite/discussions/6956#discussioncomment-3706254) or is there any pitfalls?

**Comment by dai-shi** [comment]
Thanks for opening a discussion!

I kind of knew it didn't work with vite properly, but wasn't able to find a good solution.
I wanted some official answers from Vite. But, `process.env.NODE_ENV` is unexpected, because ESM doesn't have `process`.

Jotai v1 (and current Zustand) replaces it to `process.env.NODE_ENV`.
https://github.com/pmndrs/jotai/blob/v1.13.1/rollup.config.js#L61
But, this is for "module" condition.
We can't do it for "import" condition, because it breaks in webpack.
(`typeof process !== "undefined" && process.env?.NODE_ENV` can do but then Vite doesn't work.)

This is very annoying. We can revert "module" condition, and educate Vite users to specify it, but we wouldn't be very happy.

**Reply by maxsbelt** [reply]
Thanks for detailed answer. In such case I think it makes sense to just put `jotai` in a separate chunk.

**Reply by dai-shi** [reply]
What do you mean by putting it in a separate chunk?

(I think we can revert the "module" condition or maybe add a new custom condition.)

**Reply by maxsbelt** [reply]
I mean to put this dependency in separate vendor chunk in `vite` config:

```js
defineConfig({
  build: {
    rollupOptions: {
      output: {
        manualChunks: {
          jotai: ['jotai'],
        },
      },
    },
  },
})
```

Initially I wanted to put it in the chunk with other 3rd party libraries that's updated rarely so browser can cache it and keep between deploys.

**Reply by dai-shi** [reply]
Oh, I see it's about application vite config.

**Comment by dai-shi** [comment]
> And `vite` replace it with:
> 
> ```
> { ACCESS_TOKEN: 'xxx', PROD: true, ... }&&"production"!=="production"
> ```
> 
> I think it can be problematic because if this library be extracted to `vendors` chunk it will invalidate this chunk on each env variables update which is bit unexpected.

I'm not sure if I understand the problem correctly. Can you elaborate it please?

(Actually, my concern was if Vite replaces `import.meta.env`, but if it's replaced, it sounds good to me at least in that sense.)

**Reply by maxsbelt** [reply]
Yes, sure. The issue (which is not actually an `issue` but something that can probably be improved) is in the way how `vite` replaces env variables during the build.

To reproduce you need to bootstrap new vite project and replace `main.tsx` file content with:

```js

import { createStore } from 'jotai';

const store = createStore();
console.log(store);
```

If you will build the project and check out output you will see all environment variables object at least 10 times in the output.
|output example (viewed in https://prettier.io/playground)|
|-|
|<img width="676" alt="Screenshot 2023-06-06 at 08 11 03" src="https://github.com/pmndrs/jotai/assets/950216/e984a263-0135-4ec9-bb6c-8bcfa28f45b1">|


It seems strange because the library checks only mode (production or not). If you will check the file `node_modules/jotai/esm/vanilla.mjs` it contains multiple conditions like:

```js
(import.meta.env && import.meta.env.MODE) !== "production"
```

If you will replace this conditions with:

```js
import.meta.env.MODE !== 'production'
```

It will remove all environment variables from the output and even will exclude all debug related code from the final bundle (so the final bundle will be x1.5 smaller). If you will replace this condition with:

```js
process.env.NODE_ENV !== 'production'
```

It also will exclude all environment variables but looks like it still keep some debug code, but I'm not sure.

In summary those conditions updates seems reasonable to me but I'm not sure about possible pitfalls that can happen in other environments (e.g. in webpack/parcel/etc.).

**Reply by dai-shi** [reply]
> If you will replace this conditions with:
> 
> ```js
> import.meta.env.MODE !== 'production'
> ```
> 
> It will remove all environment variables from the output

That's unexpected for me, because I expect the minification process eliminate the code in either case.

<img width="1051" alt="image" src="https://github.com/pmndrs/jotai/assets/490574/e1b559c7-0252-4bea-a8d6-649b472067f8">

Can you see why your final code is not properly minified?

**Reply by maxsbelt** [reply]
I think the problem is with the way brackets are placed. Function `f2` is from your example. Please checkout `f3` function code:

```js
export function f2() {
  (({}) && 'production' !== 'production') && console.log('debug')
  return 'hey';
}

export function f3() {
  (({} && 'production') !== 'production') && console.log('debug')
  return 'hey';
}
```

This is how it presented in the final bundle:
<img width="1054" alt="Screenshot 2023-06-06 at 12 49 06" src="https://github.com/pmndrs/jotai/assets/950216/cfad05c3-4a69-4797-bed8-24d456adddf0">

**Reply by dai-shi** [reply]
Wow, I see the issue now:
<img width="1027" alt="image" src="https://github.com/pmndrs/jotai/assets/490574/1ca492c2-3a24-4e5c-8a64-78861058a5ea">

This is _unexpected_. Thanks for the reproduction.

**Reply by dai-shi** [reply]
@maxsbelt Can you check if #1978 fixes it?
https://ci.codesandbox.io/status/pmndrs/jotai/pr/1978
Find "Local Install Instructions" ☝️

**Reply by maxsbelt** [reply]
Yeah, now it looks great 👍

**Reply by dai-shi** [reply]
Thanks for confirming. I will do the same fix in Zustand and Valtio.

**Reply by dai-shi** [reply]
@maxsbelt I just modified #1978 slightly. Can you double check if it work for you?

**Reply by maxsbelt** [reply]
@dai-shi Unfortunately no.

<img width="1227" alt="Screenshot 2023-06-07 at 20 57 14" src="https://github.com/pmndrs/jotai/assets/950216/288b6e5e-6a3c-4295-934d-9b3f8f742369">

Env variables are returned as well as debug code so result bundle size bigger:

```
// prev fix
dist/assets/index-f5b2c30f.js  4.71 kB │ gzip: 2.07 kB
// last fix
dist/assets/index-afc400ad.js  6.53 kB │ gzip: 2.45 kB
```

**Reply by dai-shi** [reply]
Oh, no! It's working with a simpler form in the playground.

<img width="1210" alt="image" src="https://github.com/pmndrs/jotai/assets/490574/12d1adc6-9eca-4d8a-a76e-e74f10ed444b">

```js
export function f3() {
  (({} && 'production') !== 'production') && console.log('debug')
  return 'hey';
}

export function f4() {
  ((!!{} && 'production') !== 'production') && console.log('debug')
  return 'hey';
}

export function f5() {
  ({} ? 'production' : undefined) !== 'production' && console.log('debug')
  return 'hey';
}
```

Maybe we want to try `f5` style?

**Reply by maxsbelt** [reply]
I tried to update imports to `(import.meta.env ? import.meta.env.MODE : undefined) !== "production"` and it works 🎉

**Reply by dai-shi** [reply]
Great to hear that! #1978 is changed too.

**Reply by maxsbelt** [reply, marked as answer]
The `v2.2.0` works perfectly. Thanks! 👍
